src/gui.cpp

```diff
@@ -1417,27 +1420,47 @@ moby_model* gui::model_browser::render_selection_grid(
 
 void gui::model_browser::render_preview(
 		app& a,
-		GLuint* target,
+		GLuint* target,	
 		moby_model& model,
 		std::vector<texture>& textures,
 		const gl_renderer& renderer,
 		ImVec2 preview_size,
 		view_params params) {
-	glm::vec3 eye = glm::vec3(2.f * (1.1f - params.zoom), 0, 0);
+
+	auto draw_list = ImGui::GetWindowDrawList();
+
+    glm::vec3 center_point = (model.bounding_box.max + model.bounding_box.min) * 0.5f / (float) INT16_MAX; 
 	
-	glm::mat4 view_fixed = glm::lookAt(eye, glm::vec3(0, 0, 0), glm::vec3(0, 1, 0));
+	//Figure out a rough camera distance 
+	float bounding_box_width = glm::abs(model.bounding_box.max.x - model.bounding_box.min.x),
+			bounding_box_height = glm::abs(model.bounding_box.max.y - model.bounding_box.min.y),
+			fov_y = 45.0f,
+			camera_distance;
+
+	if (bounding_box_width > bounding_box_height) {
+		float fov_x = fov_y * preview_size.x / preview_size.y;
+		//Get a radio of how wide the bounding box of the model is compared to the render window width.
+		float zoom_ratio = bounding_box_height / preview_size.x / (float) INT16_MAX * 10.0f;
+		//Figure out the 'default' camera depth to fit an object of width preview_size.x on screen.
+		float initial_depth = (preview_size.x/2) * tan(glm::radians(fov_x));
+		//Fit the camera to the model bounding box.
+		camera_distance = zoom_ratio * initial_depth;
+	} else {	//Same as above, but use height if the model is taller than wide.
+		float zoom_ratio = bounding_box_height / preview_size.y / (float) INT16_MAX * 10.0f;
+		float initial_depth = (preview_size.y/2) * tan(glm::radians(fov_y));
+		camera_distance = zoom_ratio * initial_depth;
+	}
+
+	glm::vec3 eye = glm::vec3(((camera_distance) * (1.1f - params.zoom)), 0, 0);
+
+	glm::mat4 view_fixed = glm::lookAt(eye, glm::vec3(0), glm::vec3(0, 1, 0));
```

Comment:
This seems to be orienting models top-down, which is a bit strange. Try testing it with ARMOR.WAD (8.WAD if you use Wrench's ISO extractor on R&C2). I'd recommend tweaking this so that models are oriented front-on instead.

![Screenshot_20210412_221938](https://user-images.githubusercontent.com/43898262/114466351-48b91200-9be0-11eb-80f4-0b8a75daaa1d.png)

Reply:
I suspect this is due to some confusion surrounding the coordinate system. The Ratchet & Clanks games all use a "Z is up" coordinate system, while OpenGL is a "Y is up" coordinate system.

Reply:
You can use the following matrix to convert between the two:

>	static const glm::mat4 yzx {
>	glm::mat4 local_to_clip = projection * offset_view;
>		0,  0, 1, 0,
>		1,  0, 0, 0,
>		0, -1, 0, 0,
>		0,  0, 0, 1
>	};

Reply:
In general I'd suggest working with the game's own coordinate system for as long as possible, then just using this matrix to convert to OpenGL coordinates at the end.

Reply:
Yep, figured out what that mat was for when you mentioned the incorrect rotation stuff. It's worth noting that this does make some of the preview windows less useful - i,e the wrenches just show the top face of the wrench, which is basically just a grey rectangle and so isn't very useful. I wonder whether doing something like orientating the camera to a non-axis aligned rotation might be the most visually useful. At any rate, I've fixed this and am working on the other change requests now :+1:

Reply:
Sounds good.